train_decoder.py

```diff
@@ -0,0 +1,462 @@
+        print(f"Recalling trainer from file: {local_filepath}")
+        state_dict = torch.load(local_filepath)
+    trainer.load_state_dict(state_dict["trainer"])
+    trainer.decoder.load_state_dict(state_dict["decoder"])
+    return trainer, state_dict["epoch"], state_dict["step"], state_dict["validation_losses"] if "validation_losses" in state_dict else []
+
+def train(
+    dataloaders,
+    decoder,
+    inference_device,
+    using_wandb=False,
+    resume_from=None,
+    resume_config=None,
+    epoch_samples = None,  # If the training dataset is resampling, we have to manually stop an epoch
+    validation_samples = None,
+    epochs = 20,
+    n_sample_images = 5,
+    save_every_n_samples = 100000,
+    amp=False,
+    lr=1e-4,
+    wd=0,
+    use_ema=True,
+    max_grad_norm=None,
+    base_path=None,
+    save_all=False,
+    save_latest=True,
+    save_best=True,
+    **kwargs
+):
+    """
+    Trains a decoder on a dataset.
+    """
+    # Create the base path if it doesn't exist
+    if not os.path.exists(base_path):
+        os.makedirs(base_path)
+    # If we are saving all, we also need to make a checkpoints folder inside there
+    if save_all:
+        if not os.path.exists(os.path.join(base_path, "checkpoints")):
+            os.makedirs(os.path.join(base_path, "checkpoints"))
+    trainer = DecoderTrainer(
+        decoder,
+        lr = lr,
+        wd = wd,
+        amp = amp,
+        use_ema = use_ema,
+        max_grad_norm = max_grad_norm
+    )
+    # Set up starting model and parameters based on a recalled state dict
+    start_step = 0
+    start_epoch = 0
+    validation_losses = []
+    if resume_from is not None:
+        if resume_from == "wandb":
+            trainer, start_epoch, start_step, validation_losses = recall_trainer(
+                trainer,
+                wandb_run_path=resume_config["wandb_run_path"],
+                wandb_file_path=resume_config["wandb_file_path"],
+            )
+        elif resume_from == "file":
+            trainer, start_epoch, start_step, validation_losses = recall_trainer(
+                trainer,
+                local_filepath=resume_config["filepath"]
+            )
+        else:
+            raise ValueError("resume_from must be either wandb or file")
+    trainer = trainer.to(device=inference_device)
+    
+    send_to_device = lambda arr: [x.to(device=inference_device, dtype=torch.float) for x in arr]
+    step = start_step
+    for epoch in range(start_epoch, epochs):
+        print(f"=========== Starting epoch {epoch} ===========")
+        decoder.train()
+        sample = 0
+        last_sample = 0
+        last_snapshot = 0
+        last_time = time.time()
+        losses = []
+        for i, (img, emb) in enumerate(dataloaders["train"]):
+            step += 1
+            sample += img.shape[0]
+            img, emb = send_to_device((img, emb))
+            
+            for unet in range(1, trainer.num_unets+1):
```

Comment:
so the other thing is, different unets in the decoder may require different amounts of training, so the researcher may have to specify which specific unet they want trained in the config

it is one of the pros of the cascading ddpm (but perhaps also the training script will be slightly more involved, as one is training all different unets to completion)

Reply:
Is isolating the unets as simply as only running the forward pass on the unets that are being trained? If that's the case I can put that in the config.

Reply:
@Veldrovive yup, you simply pass the appropriate `unet_number`  for `forward`, and then `update`

Reply:
The training configuration now has a `unet_training_mask` which is a boolean array that defines which unets will be trained.